cloudimg/ms_azure.py

```diff
@@ -0,0 +1,480 @@
+    """
+    Get all the values from the connection string.
+
+    Args:
+        connection_string (str): The connection string of the
+                                 storage account
+
+    Returns:
+        A dictionary with all the key-values in the connection string
+    """
+    res = {}
+    val_list = connection_str.split(";")
+    for val in val_list:
+        if "=" in val:
+            entry = val.split("=", 1)
+            res[entry[0]] = entry[1]
+    return res
+
+
+def get_tags_filter_expression(tags):
+    """
+    Create a filter expression from the given tags for search.
+
+    The filter expression is a string with format:
+    ```
+    "tag1"='value1' and "tag2"='value2' ...
+    ```
+
+    Args:
+        tags (dict): Dicionary of strings with the tags to filter
+
+    Returns:
+        String with the filter expression.
+    """
+
+    # Create the filter expression like:
+    # "tag1"='value1" and "tag2" = 'value2" ...
+    filter_expression = ""
+    for k, v in tags.items():
+        filter_expression += "\"%s\"='%s' and " % (k, v)
+    # strip the last ' and '
+    return filter_expression.rstrip(" and ")
+
+
+@attr.s
+class UploadProgress(object):
+    """
+    Responsible for storing the upload progress.
+
+    It's expected to be instantiated by a callback function during blob
+    upload phase.
+
+    Args:
+        current (float): The current uploaded size in bytes
+        total (float): The total size to upload in bytes
+        percentage (float, optional): The upload progress percentage
+    """
+
+    current = attr.ib(type=float, repr=lambda x: "%.2fMB" % (x / 1024**2))
+    total = attr.ib(type=float, repr=lambda x: "%.2fMB" % (x / 1024**2))
+    percentage = attr.ib(type=float, repr=lambda x: "%.2f%%" % (x))
+    last_seen = attr.ib(type=float, repr=False, eq=False, factory=time.time)
+
+    @percentage.default
+    def _auto_percentage(self):
+        if self.total > 0:
+            return self.current / self.total * 100.0
+        return 0
+
+
+class AzurePublishingMetadata(PublishingMetadata):
+    """
+    A collection of metadata necessary for uploading and publishing a
+    Virtual Hard Disk (VHD) to Azure.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(AzurePublishingMetadata, self).__init__(*args, **kwargs)
+
+        assert self.container, 'A container must be defined'
+        assert self.tags, 'A tag must be defined'
+
+
+class AzureService(BaseService):
+    """
+    Azure cloud provider service.
+    It's expected to be instanciated through one of its factories.
+
+    Args:
+        blob_service_client (BlobServiceClient): The blob service client to
+                                                 manipulate the storage.
+        storage_account_name (str): The storage account name for auth.
+        storage_account_key (str): The storage account key for auth.
+    """
+    # The duration of generated SAS URIs
+    SAS_EXPIRY = 365 * 3
```

Comment:
In cases like this I suggest that you make the magic number tunable by an environment variable, even if you don't know any specific reason it'd have to change.

It can be as simple as e.g.

```python
    SAS_EXPIRY = int(os.getenv("CLOUDIMG_AZURE_SAS_EXPIRY")) or 365*3
```

Otherwise you can find yourself in a frustrating situation where there is an issue in production, you believe it can be solved by tuning some of these numbers but you've left yourself no way to do so without a new code release (which might be blocked for any number of reasons).

It's also useful just for testing, e.g. if you wanted to try some different values for the concurrency.

Reply:
Thanks, that's a great idea!

I took advantage to do it in the other class attributes as well as we may need to fine tune these values in the future.